proposals/0023-scalar-constant-buffer-layout.md

```diff
@@ -0,0 +1,300 @@
+;       float3 Bar;                                   ; Offset:    8
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    20
+;
+; }
+```
+
+As all values must be aligned to their scalar size, a scalar or element of a
+composite type may not lay across the 16 byte boundary.
+In the following example the double is aligned identically in both legacy and
+scalar layout, to 8 bytes, demonstrating this point.
+
+```
+[bufferlayout("legacy")]
+cbuffer MyBuffer {
+  float3 Foo;
+  double Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float3 Foo;                                   ; Offset:    0
+;       double Bar;                                   ; Offset:   16
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    24
+;
+; }
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer {
+  float3 Foo;
+  double Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float3 Foo;                                   ; Offset:    0
+;       double Bar;                                   ; Offset:   16
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    24
+;
+; }
+```
+
+Matrix types in legacy layout are significantly different to scalar layout.
+With legacy layout a matrix is laid out so each row is 16 byte aligned.
+The exception to this is a matrix of doubles, in which case a double3x and
+double4x matrix will have the rows aligned to 32 bytes.
+With scalar layouts, as with vectors, each element is aligned to it's own scalar
+size.
+The following example demonstrates the differences in matrix layout by
+showing the cbuffer loads and extractvalue instructions needed to load each
+element of a 2x2 float matrix.
+Note that the legacy example requires two 16 byte loads from the cbuffer, while
+the scalar example requires only one.
+
+
+```
+[bufferlayout("legacy")]
+cbuffer MyBuffer {
+  float2x2 Foo;
+};
+
+  %loadOne = call %dx.types.CBufRet.f32 @dx.op.cbufferLoadLegacy.f32(i32 59, %dx.types.Handle %cbufferHandle, i32 0)
+  %elemOne = extractvalue %dx.types.CBufRet.f32 %loadOne, 0
+  %elemTwo = extractvalue %dx.types.CBufRet.f32 %loadOne, 1
+  %loadTwo = call %dx.types.CBufRet.f32 @dx.op.cbufferLoadLegacy.f32(i32 59, %dx.types.Handle %cbufferHandle, i32 1)
+  %elemThree = extractvalue %dx.types.CBufRet.f32 %loadTwo, 0
+  %elemFour = extractvalue %dx.types.CBufRet.f32 %loadTwo, 1
+
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer {
+  float2x2 Foo;
+};
+
+  %loadOne = call %dx.types.CBufRet.f32 @dx.op.cbufferLoadLegacy.f32(i32 59, %dx.types.Handle %cbufferHandle, i32 0)
+  %elemOne = extractvalue %dx.types.CBufRet.f32 %loadOne, 0
+  %elemTwo = extractvalue %dx.types.CBufRet.f32 %loadOne, 1
+  %elemThree = extractvalue %dx.types.CBufRet.f32 %loadThree, 2
+  %elemFour = extractvalue %dx.types.CBufRet.f32 %loadFour, 3
+
+```
+
+### SPIR-V Support
+
+SPIR-V already supports this scalar layout proposal through the
```

Comment:
nit:

```suggestion
SPIR-V already supports this scalar layout proposal, as exercised by the
```

(noting that SPIR-V can express this natively, the Vulkan extension just makes it valid to use with Vulkan)